src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs

```diff
@@ -306,15 +347,44 @@ private long Initialize()
                     {
                         count++;
 
-                        var doc = _fieldsToFetch.IsProjection
-                            ? _resultsRetriever.GetProjectionFromDocument(document, null, QueryResultRetrieverBase.ZeroScore, _fieldsToFetch, _context, null)
-                            : _inner.Current;
+                        Document doc;
+
+                        if (_fieldsToFetch.IsProjection)
+                        {
+                            var result = _resultsRetriever.GetProjectionFromDocument(document, null, QueryResultRetrieverBase.ZeroScore, _fieldsToFetch, _context, null);
+                            if (result.Document != null)
+                            {
+                                if (IsStartingPoint(result.Document))
+                                    break;
+                            }
+                            else if (result.List != null)
+                            {
+                                bool match = false;
+                                foreach (Document item in result.List)
+                                {
+                                    if (IsStartingPoint(item))
+                                    {
+                                        match = true;
+                                        break;
+                                    }
+                                }
 
-                        if (doc.Data.Count <= 0)
-                            continue;
+                                if (match)
+                                    break;
+                            }
+                        }
+                        else
+                        {
+                            if (IsStartingPoint(_inner.Current))
+                            {
+                                break;
+                            }
+                        }
 
-                        if (_alreadySeenProjections.Add(doc.DataHash) && _alreadySeenProjections.Count == _query.Start)
-                            break;
+                        bool IsStartingPoint(Document d)
+                        {
+                            return d.Data.Count > 0 && _alreadySeenProjections.Add(d.DataHash) && _alreadySeenProjections.Count == _query.Start;
```

Comment:
It's a little bit tricky that `IsStartingPoint` calls `Add()` under the covers and modifies `_alreadySeenProjections`. It's a local method so not a big issue.

Reply:
The idea was to wrap the logic to call it from both locations, yes.

Reply:
And yes, the fact that this is a local method and is right there was a factor